src/vivarium/framework/results/manager.py

```diff
@@ -1,21 +1,159 @@
-from typing import Callable, List
+from collections import Counter
+from typing import Callable, List, Union
 
 import pandas as pd
 
+# TODO: importing Builder causes circular import, fix that in implementation
+# from vivarium.framework.engine import Builder
+from vivarium.framework.event import Event
+from vivarium.framework.results.context import ResultsContext
+
 
 class ResultsManager:
+    """Backend manager object for the results management system.
+
+    The :class:`ResultManager` actually performs the actions needed to
+    stratify and observe results. It contains the public methods used by the
+    :class:`ResultsInterface` to register stratifications and observations,
+    which provide it with lists of methods to apply in their respective areas.
+    It is able to record observations at any of the time-step sub-steps
+    (`time_step__prepare`, `time_step`, `time_step__cleanup`, and
+    `collect_metrics`).
+    """
+
+    def __init__(self):
+        self._metrics = Counter()
+        self._results_context = ResultsContext()
+        self._required_columns = {"tracked"}
+        self._required_values = set()
+        self._name = "results_manager"
+
     @property
-    def name(self):
-        return "results_manager"
+    def name(self) -> str:
+        """The name of this ResultsManager."""
+        return self._name
 
-    def add_mapping_strategy(self, *args, **kwargs):
-        ...
+    @property
+    def metrics(self):
+        return self._metrics.copy()
 
-    def add_default_grouping_columns(self, *args, **kwargs):
-        ...
+    # noinspection PyAttributeOutsideInit
+    def setup(self, builder):
+        self.population_view = builder.population.get_view([])
+        self.clock = builder.time.clock()
+        self.step_size = builder.time.step_size()
+
+        builder.event.register_listener("time_step__prepare", self.on_time_step_prepare)
+        builder.event.register_listener("time_step", self.on_time_step)
+        builder.event.register_listener("time_step__cleanup", self.on_time_step_cleanup)
+        builder.event.register_listener("collect_metrics", self.on_collect_metrics)
+
+        self.get_value = builder.value.get_value
+
+        builder.value.register_value_modifier("metrics", self.get_results)
+
+    def on_time_step_prepare(self, event: Event):
+        self.gather_results("time_step__prepare", event)
+
+    def on_time_step(self, event: Event):
+        self.gather_results("time_step", event)
+
+    def on_time_step_cleanup(self, event: Event):
+        self.gather_results("time_step__cleanup", event)
+
+    def on_collect_metrics(self, event: Event):
+        self.gather_results("collect_metrics", event)
 
-    def register_results_producer(self, *args, **kwargs):
+    def gather_results(self, event_name: str, event: Event):
         ...
```

Comment:
Python doesn't error out during compilation?

Reply:
why would it?

Reply:
TIL: you can use the ellipsis object in place of pass